### PR TITLE
[MRG] implement padding for max pool grad grad

### PR DESCRIPTION
**Pull request by yaoli**
max pool grad grad python code added

all tests passed

### CODE REVIEW

File: theano/tensor/signal/downsample.py [hunk 1]
```diff
@@ -788,26 +808,38 @@ def make_node(self, x, maxout, gz):
...
         st0, st1 = self.st
-        img_rows = x.shape[-2]
-        img_cols = x.shape[-1]
+        pd0, pd1 = self.padding
+        img_rows = x.shape[-2] + 2 * pd0 
+        img_cols = x.shape[-1] + 2 * pd1
 
+        # pad the image and its gradients
+        if self.padding != (0, 0):
+            y_padded = numpy.zeros(
```
Comment (nouiz) [inline theano/tensor/signal/downsample.py]: can you rename to x_padded for consistency in that method?

File: theano/tensor/signal/downsample.py [hunk 2]
```diff
@@ -818,8 +850,8 @@ def perform(self, node, inp, out):
                         col_end = __builtin__.min(col_st + ds1, img_cols)
                         for row_ind in xrange(row_st, row_end):
                             for col_ind in xrange(col_st, col_end):
-                                if (maxout[n, k, r, c] == x[n, k, row_ind, col_ind]):
-                                    ggz[n, k, r, c] = ggx[n, k, row_ind, col_ind]
-
+                                if (maxout[n, k, r, c] == y_padded[n, k, row_ind, col_ind]):
+                                    ggz[n, k, r, c] = ggx_padded[n, k, row_ind, col_ind]
+
```
Comment (nouiz) [inline theano/tensor/signal/downsample.py]: pep8

File: theano/tensor/signal/downsample.py [hunk 3]
```diff
@@ -694,10 +694,10 @@ def c_code_cache_version(self):
         return (0, 7)
 
 class DownsampleFactorMaxGradGrad(Op):
-    __props__ = ('ds', 'ignore_border', 'st')
+    __props__ = ('ds', 'ignore_border', 'st', 'padding')
 
     @staticmethod
-    def out_shape(imgshape, ds, ignore_border=False, st=None):
+    def out_shape(imgshape, ds, ignore_border=False, st=None, padding=(0, 0)):
```
Comment (nouiz) [inline theano/tensor/signal/downsample.py]: This method seem a copy of DownsampleFactorMax.out_shape. Why not call it directly? Both of them are static method, so this is easy and remove duplicate code.

**Comment by nouiz** [comment]
I did some small comments. Otherwise it look good and it is ready to get its c code implemented.

**Comment by yaoli** [comment]
just commit the c code for the grad of grad

**Comment by nouiz** [comment]
Closed and reopened to make travis rerun on the current master.

**Comment by nouiz** [comment]
The PR seem good except for my small comment. We can merge after the python23 PR is merged.
One questions, the gradgrad only support the mode max. Is it useful that you implement the other mode?

**Comment by yaoli** [comment]
I think whoever started the other mode would be more informed to finish it.
I've spent too much time on pooling already, would prefer something
different for the next ticket, for example, more gpu and c code related.
2015-6-22 上午10:35于 "Frédéric Bastien" <notifications@github.com>写道：

**Pull request merged**